### ISSUE TITLE
timesyncd allows time-sync.target to be reached before time is synchronized.

### ISSUE THREAD

**Issue by LukeShu** [systemd/systemd#5097]
### Submission type

  - [x] Bug report
  - [ ] Request for enhancement (RFE)

### systemd version the issue has been seen with

>  232-6.parabola1

### Used distribution

> Parabola GNU/Linux-libre (based on Arch Linux)

### In case of bug report: Expected behaviour you didn't see

> As systemd.special(7) says " All services where correct time is essential should be ordered after this unit [time-sync.target]", I would expect that if systemd-timesyncd is enabled, it would prevent time-sync.target from being reached until the time is synchronized and the clock is accurate.

### In case of bug report: Unexpected behaviour you saw

> timesyncd started, notified that it started; causing the system to reach time-sync.target (allowing a service ordered after-time-sync.target to start). A full 29 seconds later, the timesyncd actually managed to synchronize with the NTP server, setting the clock back about an hour (causing the service ordered after time-sync.target to fail to work for an hour until the initial time was reached[1]).

> ([1]: It then proceeded to spit an error message in the journal roughly every 5 seconds for that hour; interestingly, according to the journal, the first such error message happens 3 seconds *before* the clock jumps back.)

**Comment by LukeShu** [comment]
I've written a hacky little workaround: https://git.lukeshu.com/systemd-timesyncd-wait/

The gist is that a separate service (`systemd-timesyncd-wait.service`) orders itself Before=time-sync.target, and waits for a notification from systemd-timesyncd that it has synchronized.

I avoided simply having systemd-timesyncd avoid sending READY=1 until it has synced because (1) I was worried about the unit timing out, and (2) it is ordered Before=sysinit.target, which I didn't want to hold up.  I'm not sure why it's important for it to be before sysinit.target, but I trust that whoever did that had a good reason.

If this type of solution is agreeable to be merged, the systemd-timesyncd-wrap program would be unnecessary; systemd-timesyncd itself would be patched.  The wrap program sits between timesyncd and systemd, inspecting the sd_notify STATUS= messages from timesyncd, forwarding them to systemd, and sending a notification to systemd-timesyncd-wait when appropriate.  If this were mainlined, timesyncd would just send the notification to systemd-timesyncd-wait itself.

**Comment by poettering** [comment]
Yeah, we should probably provide a systemd-timesync-wait.service out of the box, for those who want this.

Normally though I doubt this is desirable. Delaying the boot depending on external network conditions sounds like a bad idea in normal operation of generic systems. It's only a good idea in very few, very specific setups.

**Comment by ojura** [comment]
Hey @LukeShu, your unit file did not work on Ubuntu Xenial because rm and systemd-timesyncd are not inside /usr, I had to adjust this in the .conf file (see https://github.com/assisi/systemd-timesyncd-wait).

Edit: updated repo url.

**Comment by rfrail3** [comment]
This functionality would be very helpful. Thanks

**Comment by LukeShu** [comment]
> Delaying the boot depending on external network conditions sounds like a bad idea in normal operation of generic systems. It's only a good idea in very few, very specific setups.

While generally I agree, that *is* what `time-sync.target` is for; if you are ordered after it it means that you are willfully asking to be ordered after the system has established communication "from a remote source". The facility shouldn't be broken just because using the facility is usually a bad choice.

**Comment by tkfu** [comment]
I'd just like to chime in to say that there are definitely use cases where this is essential. The one I ran into today is this: you have an image that includes some device provisioning key which you flash on a bunch of embedded devices. The first time they boot, they contact the provisioning server, present the provisioning key, and get individual device credentials. If time isn't yet synced, the device's time may be outside of the validity of the provisioning server's certificate, causing it to fail provisioning.

I've worked around it for my particular application, but I will say that I'd agree with @LukeShu that `time-sync.target` is highly misleading if it doesn't actually deliver synced time--this should be listed as a bug, not an RFE.

**Comment by karlito-go** [comment]
So I run systemctl show time-sync.target:
    
      Description=System Time Synchronized 

I look at my boot messages with microsecond timestamps: journalctl -b -o  short-precise:

    Aug 01 20:52:16.351608 test2 systemd[1]: Reached target System Time Synchronized.
    Aug 01 20:52:16.361146 test2 systemd[1]: Starting Raise network interfaces...
    ... Aug 01 20:52:16.402797 test2 dhclient[710]: DHCPDISCOVER on eth0 to 255.255.255.255 port 67 interval 3 (xid=0x90ab5330)

Huh.  System time is synchronized before network/dhcp even starts?  Wat?

I look at the source code: https://github.com/systemd/systemd/blob/master/src/timesync/timesyncd.c#L143

Huh, it sends sd_notify "READY=1\n" as soon as it... reads its config file?  Wat?

We run AWS machines that generate type 1 (time) UUIDs.  We would really like time to be synchronized before things start.  I didn't realize that was a "specific setup".

The documentation for timesyncd and these targets is completely misleading.  Does it even make the "no backwards time" guarantee before saying "ready!"?

**Comment by LukeShu** [comment]
I've adjusted `systemd-timesyncd-wait` to work correctly with systemd 234's unit files (it was broken by the default `systemd-timesyncd.service` changing from `ProtectSystem=full` to `ProtectSystem=strict`), as well as merging @ojura's changes to make it Ubuntu-friendly, and tidying up a bit.

I also made a GitHub mirror of it.
https://git.lukeshu.com/systemd-timesyncd-wait/
https://github.com/LukeShu/systemd-timesyncd-wait

I suppose that at this point, the only way for this bug (and yes, it is a bug, not on RFE) to get fixed is to rewrite systemd-timesyncd-wait in C, and put it in a pull request.

**Comment by LukeShu** [comment]
Actually, I did implement the `-wait` process in C already https://lists.freedesktop.org/archives/systemd-devel/2017-May/038961.html ; all that remains is patching systemd-timesyncd to talk to it (rather than using a wrapper).

**Comment by poettering** [comment]
Yeah, to be merged into systemd we need it in C. Happy to take a patch for that

**Comment by albundy83** [comment]
Hello,
under raspberry pi 3, we use raspbian which is based from debian 8 jessie and systemd is older than this version (systemd version 215), do you have an idea on how make it works this great go tool ?
I have open an issue : https://github.com/assisi/systemd-timesyncd-wait/issues/2

**Comment by SimonvanEeden** [comment]
I was trying to install this on my raspberry pi 3. I coud not get it to work and i noticed the service was getting installed at usr//lib/systemd... instead of /lib/systemd...
I wanted to uninstall the old files but i did nit know how, i moved files over to /lib/systemd.. but now my time went back to 2016 and I think there is a whole lot going wrong.
What do i do now, i just want my time back.

I had also changed a config file to point to the moved files.

**Comment by albundy83** [comment]
Hello,
depending on where you have installed this great tool, first stop systemd-timesyncd-wait service
```bash
systemctl stop systemd-timesyncd-wait
```
Then clean install:
```bash
rm -f /usr/local/lib/systemd/systemd-timesyncd-wait
rm -f /usr/local/lib/systemd/systemd-timesyncd-wrap
rm -f /usr/local/lib/systemd/system/systemd-timesyncd-wait.socket
rm -f /usr/local/lib/systemd/system/systemd-timesyncd-wait.service
rm -rf /usr/local/lib/systemd/system/systemd-timesyncd.service.d/
```
or
```bash
rm -f /usr/lib/systemd/systemd-timesyncd-wait
rm -f /usr/lib/systemd/systemd-timesyncd-wrap
rm -f /usr/lib/systemd/system/systemd-timesyncd-wait.socket
rm -f /usr/lib/systemd/system/systemd-timesyncd-wait.service
rm -rf /usr/lib/systemd/system/systemd-timesyncd.service.d/
```
Once it's cleaned, just issue
```bash
systemctl daemon-reload
systemctl restart systemd-timesyncd
```
This will update systemd-timesyncd configuration and restart it.

**Comment by albundy83** [comment]
Be careful also as this tool is not compatible with old raspbian release (jessie based one).
The newest raspbian based on stretch is ok.
To check, just issue
```bash
systemd --version
```
You must be at least at 229 something
With latest raspbian stretch, you will be at version 232
Use this fork instead
https://github.com/LukeShu/systemd-timesyncd-wait

**Comment by SimonvanEeden** [comment]
Thanks, I got it working.
I'm using Raspbian on the PI3 version 232, the command that worked for me was "sudo make prefix= install"
This way it installed in /lib/systemd/.... just like my other homemade service and it did work.
Getting go commands to work with sudo didn't work but then it suddenly did :)

**Comment by SimonvanEeden** [comment]
A new issue appeared, when i add my own NTP server in /etc/timesyncd.conf
systemctl status systemd-timesyncd gives: https://pastebin.com/6pyNwcA1

when i then reload the deamon only the warning goes
when i then restart systemd-timesyncd: https://pastebin.com/quXUH2Mi

the wait service seems to have blocked timesyncd to early.
Edit: 
needed to restart my pc for ntp firewall settings to change *facepalm*

**Comment by BeardOverflow** [comment]
Thanks to all for working in solve this problem.
Without using systemd-timesyncd-wait, a possible workaround would be add in .service file:

ExecStartPre=/usr/sbin/ntp-wait

**Comment by karniemi** [comment]
Another angle on this issue...for virtualization:
* at least one (major) commercial virtualization product recommends using ntp to sync time in guest systems, instead of syncing with the host. However - they still recommend to sync with host (using the products guest tools) for one time events like "startup, taking a snapshot, resuming from a snapshot, resuming from suspend".  Which of course implies that they also recommend keeping host properly synced with NTP or some other mechanism.
* for a system that follows that recommendation, I would like to imply that "time-sync.target" is reached in VMs when either 1) the time was synced from the host or  2) when ntp sync is complete ... which ever comes first.
  * I could not figure out the exact point in guest boot process for the one time sync. Does it happen somehow before boot  - or during the guest as part of their guest tools. And if it's the latter...how much does it conflict with the ntpd performing sync in parallel...

**Comment by GeekGoutham** [comment]
@LukeShu Were you able to merge the patch for this issue? The patch can really help me avoid writing additional systemd units to force time sync before other packaged open source systemd units start.

**Comment by pabigot** [comment]
I believe this should now be resolved with PR #8494.  With the merged solution when `systemd-time-wait-sync.service` is enabled `time-sync.target` will be blocked until the adjtimex(2) status from the kernel indicates that the clock is synchronized.  This works with both timesyncd from systemd and with connman's internal NTP client.  It might not work with a time synchronization solution that doesn't follow the NTP protocol for setting the realtime clock in the kernel.

Example below.  The first timestamp is from the file system time; the second from the timestamp restored by timesyncd when it starts; the final one when timesyncd sets the system clock and marks it synchronized.

```
root@raspberrypi3:~# systemctl status systemd-timesyncd
��● systemd-timesyncd.service - Network Time Synchronization
   Loaded: loaded (/lib/systemd/system/systemd-timesyncd.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2018-03-20 22:45:51 UTC; 14h ago
     Docs: man:systemd-timesyncd.service(8)
 Main PID: 147 (systemd-timesyn)
   Status: "Synchronized to time server 216.239.35.0:123 (time1.google.com)."
   CGroup: /system.slice/systemd-timesyncd.service
           ��└��─147 /lib/systemd/systemd-timesyncd

Mar 20 22:27:28 raspberrypi3 systemd[1]: Starting Network Time Synchronization...
Mar 20 22:27:28 raspberrypi3 systemd-timesyncd[147]: System clock time unset or jumped backwards, restoring from recorded timestamp: Tue 2018-03-20 22:45:51 UTC
Mar 20 22:45:51 raspberrypi3 systemd[1]: Started Network Time Synchronization.
Mar 20 22:46:03 raspberrypi3 systemd-timesyncd[147]: Network configuration changed, trying to establish connection.
Mar 21 13:03:01 raspberrypi3 systemd-timesyncd[147]: Synchronized to time server 216.239.35.0:123 (time1.google.com).
root@raspberrypi3:~# systemctl status systemd-time-wait-sync                                                                                                             
��● systemd-time-wait-sync.service - Wait Until Kernel Time Synchronized
   Loaded: loaded (/lib/systemd/system/systemd-time-wait-sync.service; enabled; vendor preset: enabled)
   Active: active (exited) since Wed 2018-03-21 13:03:01 UTC; 19s ago
     Docs: man:systemd-time-wait-sync.service(8)
  Process: 82 ExecStart=/lib/systemd/systemd-time-wait-sync (code=exited, status=0/SUCCESS)
 Main PID: 82 (code=exited, status=0/SUCCESS)

Mar 20 22:27:24 raspberrypi3 systemd-time-wait-sync[82]: adjtime state 5 status 40 time Tue 2018-03-20 22:27:23.977921 UTC
Mar 20 22:45:51 raspberrypi3 systemd-time-wait-sync[82]: adjtime state 5 status 40 time Tue 2018-03-20 22:45:51.265334 UTC
Mar 21 13:03:01 raspberrypi3 systemd-time-wait-sync[82]: adjtime state 0 status 2000 time Wed 2018-03-21 13:03:01.215343 UTC
Mar 21 13:03:01 raspberrypi3 systemd[1]: Started Wait Until Kernel Time Synchronized.
Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.
root@raspberrypi3:~# systemctl status time-sync.target
��● time-sync.target - System Time Synchronized
   Loaded: loaded (/lib/systemd/system/time-sync.target; static; vendor preset: enabled)
   Active: active since Wed 2018-03-21 13:03:01 UTC; 26s ago
     Docs: man:systemd.special(7)

Mar 21 13:03:01 raspberrypi3 systemd[1]: Reached target System Time Synchronized.
```

**Comment by fsateler** [comment]
> I believe this should now be resolved with PR #8494.

Indeed, it is. Closing now.

Next time, you can include a reference: Fixes: #NNN and have the issue be automatically closed.

**Comment by hacker1024** [comment]
For those on NixOS, this has been discussed [here](https://github.com/NixOS/nixpkgs/pull/51338/commits/d25378a83101c6b70183ceba7bb29381ce463945).
